compiler/rustc_resolve/src/diagnostics.rs

```diff
@@ -1804,6 +1806,81 @@ impl<'a> Resolver<'a> {
         }
     }
 
+    fn find_similarly_named_type(
+        &mut self,
+        ident: Symbol,
+        ribs: Option<&PerNS<Vec<Rib<'a>>>>,
+    ) -> Option<TypoSuggestion> {
+        fn is_type_candidate(res: Res) -> bool {
+            matches!(
+                res,
+                Res::Def(
+                    DefKind::Struct
+                        | DefKind::Union
+                        | DefKind::Enum
+                        | DefKind::Trait
+                        | DefKind::TraitAlias
+                        | DefKind::TyAlias
+                        | DefKind::AssocTy
+                        | DefKind::TyParam
+                        | DefKind::OpaqueTy
+                        | DefKind::ForeignTy,
+                    _
+                ) | Res::PrimTy(..)
+                    | Res::SelfTy { .. }
+            )
+        }
+
+        let mut names = Vec::new();
+        if let Some(ribs) = ribs {
+            // Search in lexical scope.
+            // Walk backwards up the ribs in scope and collect candidates.
+            for rib in ribs[TypeNS].iter().rev() {
```

Comment:
I implemented the logics after this line with reference to the following function:
https://github.com/rust-lang/rust/blob/e013f9e0ca4055930f931a30171c355de42f9f92/compiler/rustc_resolve/src/late/diagnostics.rs#L1343